File: Reverse_Linked_List.java
/**
 * Problem: https://leetcode.com/problems/reverse-linked-list/
 * 
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode rev = null;
        while(head != null) {
            ListNode tail = head.next;
            head.next = rev;
            rev = head;
            head = tail;
        }
        return rev;
    }
    
    public ListNode reverseList(ListNode head) {
        return reverseList(head, null);
    }
    
    private ListNode reverseList(ListNode head, ListNode revHead) {
        if (head == null) return revHead;
        ListNode rest = head.next;
        head.next = revHead;
        return reverseList(rest, head);
    }
}

File: Swift/reverse-linked-list.swift
/**
 * https://leetcode.com/problems/reverse-linked-list/
 * 
 * 
 */ 
// Date: Mon May  4 22:49:29 PDT 2020
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class Solution {
    /// iterative way to solve this problem.
    func reverseList(_ head: ListNode?) -> ListNode? {
        var head = head
        var next = head?.next
        head?.next = nil
        while next != nil {
            let nnext = next?.next
            next?.next = head
            head = next
            next = nnext
        }
        return head
    }
}

class Solution {
    /// Recursive way to solve this problem.
    func reverseList(_ head: ListNode?) -> ListNode? {
        var head = head
        guard var next = head?.next else { return head }
        head?.next = nil
        let rvNode = reverseList(next)
        next.next = head
        return rvNode
    }
}
